Add the input subsystem

Cautreoxit opened the pull request:
This PR is associated with https://github.com/asterinas/asterinas/issues/2292, consising of four commits.

TinaZhangZW reviewed (commented):
Thanks for your work. Please see my comments inline.

lrh2000 commented on kernel/comps/input/src/input_handler.rs:
```diff
@@ -0,0 +1,33 @@
+
+use crate::{InputDevice, InputEvent};
+
+/// Handler class trait - represents a type of input handler (e.g., evdev, kbd)
```
https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md#summary-sentence
 - In API documentation, the first line should be a single-line short sentence providing a summary of the code.
 - The summary line should be written in third person singular present indicative form.

In addition, quote the Rust code (e.g., `Ok()`, `Err()`) in ` `` ` (i.e., `` `Ok()` ``, `` `Err()` ``) and include the final period.

I give your an example, please do fix all your doc comments, thanks.
```rust
/// A trait that represents an input handler class.
/// 
/// Once registered to the input core, the input handler class will try to
/// connect to every input device (via [`Self::connect`]). If it succeeds, an
/// [`InputHandler`] will be created to handle the input events come from that
/// device.
pub trait InputHandlerClass: Send + Sync + Any + Debug {
    /// Returns the class name of the handler class.
    fn name(&self) -> &str;

    /// Tries to connect to the input device.
    ///
    /// On success, this method will return `Ok()` with a new input handler.
    /// Otherwise, it will return `Err(ConnectError)`.
    fn connect(&self, dev: Arc<dyn InputDevice>) -> Result<Arc<dyn InputHandler>, ConnectError>;
}
```

Cautreoxit commented on kernel/comps/input/src/event_type_codes.rs:
```diff
@@ -0,0 +1,349 @@
+        const FF_STATUS = 1 << 0x17;
+    }
+}
+
```
The reason for using `bitflags!` here is that an input device’s capabilities may have multiple bits set to 1. If we were to use an `enum` instead of `bitflags!`, representing these capabilities would waste storage space, so we chose `bitflags!`. In addition, the implementation of the as_u16 method was indeed not very elegant, and I have already revised it.

Cautreoxit commented on kernel/comps/keyboard/src/i8042_chip/keyboard.rs:
```diff
@@ -56,176 +63,324 @@ pub(super) fn init(controller: &mut I8042Controller) -> Result<(), I8042Controll
-    fn is_extension(&self) -> bool {
-        self.0 == 0xE0
+/// Map ScanCode to InputKey
+fn scancode_to_input_key(keyboard_state: &KeyboardState) -> Option<InputKey> {
```
Yes, the current role of this callback function is to let the TTY map `KeyCode` to xterm escape sequences and output characters. After implementing the input subsystem, the TTY can provide another input handler besides `evdev`. In this way, the keyboard driver only needs to deliver the `KeyCode` enum, while the `evdev` handler will convert the `KeyCode` into a `KeyEvent` and pass it to user programs, and the TTY handler will convert the `KeyCode` into xterm escape sequences and output them. Since the TTY output path is not closely tied to this PR, I plan to leave this improvement for a subsequent PR.

Cautreoxit replied:
Thanks for your comments — they’re both useful and reasonable! I followed your advice and replaced the evdev handler with a tty handler. This change really helps us cut down redundant code. For instance, in kernel/comps/keyboard/src/lib.rs, we no longer need the odd InputKey type or its callback.

Regarding your additional point: the current implementation of tracking control states originally came from the callback mechanism. This is indeed incorrect for the input subsystem, since state tracking is the responsibility of higher layers, not the driver. With the tty handler in place, this design can be removed as well. For LEDs, the input subsystem handles them via a dedicated callback, which higher layers (such as the tty handler) use to notify the driver when the LED state should change. It is not supported currently, but it would not conflict with this framework.

Cautreoxit replied:
In the Linux input subsystem, LED control is handled via the event callback of `input_dev`. For example, in the PS/2 keyboard driver (`drivers/input/keyboard/atkbd.c`), the atkbd_event function processes `EV_LED` events to update hardware LEDs. The upper layer (e.g., TTY) triggers these events using input_event when the LED state changes (e.g., Caps Lock toggled). 

The link of the callback is https://github.com/torvalds/linux/blob/097a6c336d0080725c626fda118ecfec448acd0f/drivers/input/keyboard/atkbd.c#L679C1-L707C2

Cautreoxit commented:
> Thanks for your contribution. This is great work.
> 
> I reviewed commit "Migrate i8042 keyboard to input subsystem API" and went through commit "Add input subsystem". For those two commits, see comments below.
> 
> The commit "Add evdev handler" saves `FsResolver` in a global variable and using it to create device nodes, which does not seem to be a correct design. After the system has successfully booted, the userspace program (e.g., `systemd-udev`) should create the device nodes. However, some device nodes, such as those of certain block devices (e.g., `vda`), are generated in `devtmpfs` at boot time. It is unclear whether this applies to input device nodes. Even if so, populating new device nodes at runtime is still incorrect.
> 
> Given that, I didn't review the details of the evdev commit. I think it can be moved to a separate PR. Its design will rely on the design of our `devtmpfs`/`sysfs`/`uevent` systems. But as I said in [#2364 (comment)](https://github.com/asterinas/asterinas/pull/2364#discussion_r2318545783), the TTY is already a good candidate for the first user of the input subsystem. Therefore, it would make sense for the TTY to use the input subsystem in this PR, which seems less controversial than the evdev stuffs.

Thanks for your thoughtful advice! I’ve made two sets of changes based on your comments:

1. Replaced the evdev handler with a tty handler. I’ll leave the evdev implementation for a future PR. The tty handler indeed makes the code cleaner and more correct.
2. Fixed the smaller issues you pointed out, such as comment style and the race condition.

Please let me know if you spot anything else. Thanks again for your helpful feedback!

lrh2000 reviewed (commented):
Thanks for addressing the previous comments!

farooqkz commented:
I wanted to say I'm quite excited about the progress here!

Cautreoxit commented on kernel/src/device/tty/input_handler.rs:
```diff
@@ -0,0 +1,249 @@
+            return Err(ConnectError::IncompatibleDevice);
+        }
+
+        let tty = super::system_console();
```
Yes, that's right. Keys from the keyboard should not appear in the non-framebuffer console. I have fixed this error. Now the keys would only appear in the framebuffer console.

lrh2000 reviewed (commented):
Thanks for the updates. The code is getting better, while I found that some previous comments were not handled correctly or completely, plus there is a set of minor comments.

lrh2000 reviewed (commented):
Thank you for your patience in improving the code, I really appreciate it!

This is mostly a follow-up to the previous review round. I expect we can reach consensus on this PR soon.
